components/script/dom/webgl/webglprogram.rs

```diff
@@ -3,41 +3,179 @@
  * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
 
 // https://www.khronos.org/registry/webgl/specs/latest/1.0/webgl.idl
-use std::cell::Cell;
+use std::cell::{Cell, RefCell};
 use std::collections::HashSet;
 
 use canvas_traits::webgl::{
     ActiveAttribInfo, ActiveUniformBlockInfo, ActiveUniformInfo, WebGLCommand, WebGLError,
     WebGLProgramId, WebGLResult, webgl_channel,
 };
 use dom_struct::dom_struct;
+use script_bindings::weakref::WeakRef;
 
 use crate::dom::bindings::cell::{DomRefCell, Ref};
 use crate::dom::bindings::codegen::Bindings::WebGL2RenderingContextBinding::WebGL2RenderingContextConstants as constants2;
 use crate::dom::bindings::codegen::Bindings::WebGLRenderingContextBinding::WebGLRenderingContextConstants as constants;
 use crate::dom::bindings::inheritance::Castable;
 use crate::dom::bindings::reflector::{DomGlobal, reflect_dom_object};
-use crate::dom::bindings::root::{DomRoot, MutNullableDom};
+use crate::dom::bindings::root::DomRoot;
 use crate::dom::bindings::str::DOMString;
 use crate::dom::webgl::webglactiveinfo::WebGLActiveInfo;
 use crate::dom::webgl::webglobject::WebGLObject;
 use crate::dom::webgl::webglrenderingcontext::{Operation, WebGLRenderingContext};
 use crate::dom::webgl::webglshader::WebGLShader;
 use crate::dom::webgl::webgluniformlocation::WebGLUniformLocation;
+use crate::dom::webglrenderingcontext::capture_webgl_backtrace;
 use crate::script_runtime::CanGc;
 
-#[dom_struct]
-pub(crate) struct WebGLProgram {
-    webgl_object: WebGLObject,
+#[derive(JSTraceable, MallocSizeOf)]
+struct DroppableWebGLProgram {
     #[no_trace]
     id: WebGLProgramId,
     is_in_use: Cell<bool>,
     marked_for_deletion: Cell<bool>,
+    fragment_shader: RefCell<Option<WeakRef<WebGLShader>>>,
+    vertex_shader: RefCell<Option<WeakRef<WebGLShader>>>,
+    context: WeakRef<WebGLRenderingContext>,
+}
+
+impl DroppableWebGLProgram {
+    fn new(
+        id: WebGLProgramId,
+        is_in_use: Cell<bool>,
+        marked_for_deletion: Cell<bool>,
+        fragment_shader: RefCell<Option<WeakRef<WebGLShader>>>,
+        vertex_shader: RefCell<Option<WeakRef<WebGLShader>>>,
+        context: WeakRef<WebGLRenderingContext>,
+    ) -> Self {
+        Self {
+            id,
+            is_in_use,
+            marked_for_deletion,
+            fragment_shader,
+            vertex_shader,
+            context,
+        }
+    }
+}
+
+impl DroppableWebGLProgram {
+    pub(crate) fn id(&self) -> WebGLProgramId {
```

Comment:
I believe these methods are only called inside of this module. Let's remove the pub(crate) unless it's necessary.